### DISCUSSION TITLE
Linker errors - undefined ref

### DISCUSSION THREAD

**Discussion by bluedalmatian** [resiprocate/resiprocate#345] [Q&A] [answered]
Hi Im having difficulty linking against the library. ld is giving the error:
.../libresip.a(Headers.o): in function `resip::Headers::getType(char const*, int)':
..../resiprocate-1.12.0/resip/stack/Headers.cxx:350: undefined reference to `resip::HeaderHash::in_word_set(char const*, unsigned long)'

The linker options are being passed thus:
 -L ../lib/resiprocate_build/usr/local/lib -l:libresipares.a -l:libdum.a -l:libresip.a -l:librutil.a

All the libs are in that directory. Ive tried re-ordering the order they are passed but to no avail. Its a long while since Ive used resiprocate but my project used to link against version 1.4 without issue so Im obviously missing something now im trying to use 1.12?

This is with GNU ld 2.40 on Linux

Thanks

**Comment by gjasny** [comment]
Hello,
the actual function signature in the header is:
```
resip/stack/HeaderHash.hxx:  static const struct headers *in_word_set (const char *str, GPERF_SIZE_TYPE len);
```

My guess is that the two compilation units use different typedefs for `GPERF_SIZE_TYPE`. I'd use `nm` to inspect the provided and needed symbols.

The implementation provided in `HeaderHash.cxx` is either provided in the source tree if no gperf is available or (re)generated if no gperf could be found. I remember the autotools logic was a little bit icky if not outright broken for some corner cases.

If you feel uncomfortable touching autotools file, you could give the master branch with its CMake build system a try. To just compile the stack, you could find some good initial CMake variable settings [here](https://github.com/conan-io/conan-center-index/pull/20465/files).

**Comment by bluedalmatian** [comment]
It appears the HeaderHash.cxx file does not exist. Quite why is isnt being generated Im not sure.

**Comment by bluedalmatian** [comment]
Managed to get HeaderHash.cxx gernated by running gperf manually. HeaderHash.o is now in libresip.a but still same error running nm reveals no in_word_set in the resip::HeaderHash namespace:

 nm ./libresip.a -C | grep 'in_word_set'
                               U resip::DayOfWeekHash::in_word_set(char const*, unsigned long)
                               U resip::MonthHash::in_word_set(char const*, unsigned long)
0000000000000000 t _GLOBAL__sub_I__ZN5resip13DayOfWeekHash11in_word_setEPKcm
0000000000000000 T resip::DayOfWeekHash::in_word_set(char const*, unsigned long)
0000000000000000 r resip::DayOfWeekHash::in_word_set(char const*, unsigned long)::lookup
0000000000000000 d resip::DayOfWeekHash::in_word_set(char const*, unsigned long)::wordlist
                                 U resip::HeaderHash::in_word_set(char const*, unsigned long)
0000000000000000 T resip::MethodHash::in_word_set(char const*, unsigned long)
0000000000000020 r resip::MethodHash::in_word_set(char const*, unsigned long)::lookup
0000000000000000 d resip::MethodHash::in_word_set(char const*, unsigned long)::wordlist
                 U resip::MethodHash::in_word_set(char const*, unsigned long)
0000000000000000 t _GLOBAL__sub_I__ZN5resip9MonthHash11in_word_setEPKcm
0000000000000000 T resip::MonthHash::in_word_set(char const*, unsigned long)
0000000000000000 r resip::MonthHash::in_word_set(char const*, unsigned long)::lookup
0000000000000000 d resip::MonthHash::in_word_set(char const*, unsigned long)::wordlist
0000000000000000 t _GLOBAL__sub_I__ZN5resip13ParameterHash11in_word_setEPKcm
0000000000000000 T resip::ParameterHash::in_word_set(char const*, unsigned long)
0000000000000000 r resip::ParameterHash::in_word_set(char const*, unsigned long)::lookup
0000000000000000 d resip::ParameterHash::in_word_set(char const*, unsigned long)::wordlist
                                 U resip::ParameterHash::in_word_set(char const*, unsigned long)

**Comment by bluedalmatian** [comment]
oK So for some reason problem one appears to be the cxx not being generated by gperf so i ran it manually to create HeaderHash.cxx

problem two is then that HeaderHash.cxx is never compiled into HeaderHash.o so i ran g++ on it manually

problem three is that although having done that, and rebuilt the lib, nm shows the function in libresip.a and libresip.so, BUT  it lists is as 'Undefined':

 U resip::HeaderHash::in_word_set(char const*, unsigned long)

Im now at a loss.

**Comment by sgodin** [comment, marked as answer]
I remember the old build system used to run a search and replace on
the gperf generated file to change a few things.  I'm not really familiar
with the CMake build yet.  I see a file called gperf_w32 that contains how
to build the files on windows, you could extract those exact commands for
linux.

I think there may be an easier solution here though.  There are pre-gen'd
gperf files in resip/stack/gen - hopefully you can just copy them back a
directory.

Scott

**Comment by portsip** [comment]
We used the following command to compile the master branch on Linux is working fine:


cmake -G "Unix Makefiles" -DCMAKE_BUILD_TYPE=RELWITHDEBINFO -DBUILD_REPRO=FALSE -DBUILD_DSO_PLUGINS=FALSE -DBUILD_RETURN=FALSE -DBUILD_QPID_PROTON=FALSE -DREGENERATE_MEDIA_SAMPLES=FALSE -DENABLE_LOG_REPOSITORY_DETAILS=FALSE -DVERSIONED_SONAME=FALSE -DBUILD_TFM=FALSE -DBUILD_SHARED_LIBS=OFF -DCMAKE_POSITION_INDEPENDENT_CODE=ON ../../
